[SPARK-22147][CORE] Removed redundant allocations from BlockId

superbobry opened the pull request:
## What changes were proposed in this pull request?

Prior to this commit BlockId.hashCode and BlockId.equals were defined
in terms of BlockId.name. This allowed the subclasses to be concise and
enforced BlockId.name as a single unique identifier for a block. All
subclasses override BlockId.name with an expression involving an
allocation of StringBuilder and ultimatelly String. This is suboptimal
since it induced unnecessary GC pressure on the dirver, see
BlockManagerMasterEndpoint.

The commit removes the definition of hashCode and equals from the base
class. No other change is necessary since all subclasses are in fact
case classes and therefore have auto-generated hashCode and equals. No
change of behaviour is expected.

Sidenote: you might be wondering, why did the subclasses use the base
implementation and the auto-generated one? Apparently, this behaviour
is documented in the spec. See this SO answer for details
https://stackoverflow.com/a/44990210/262432.

## How was this patch tested?

BlockIdSuite

Commit:
[SPARK-22147][CORE] Removed redundant allocations from BlockId

Prior to this commit BlockId.hashCode and BlockId.equals were defined
in terms of BlockId.name. This allowed the subclasses to be concise and
enforced BlockId.name as a single unique identifier for a block. All
subclasses override BlockId.name with an expression involving an
allocation of StringBuilder and ultimatelly String. This is suboptimal
since it induced unnecessary GC pressure on the dirver, see
BlockManagerMasterEndpoint.

The commit removes the definition of hashCode and equals from the base
class. No other change is necessary since all subclasses are in fact
case classes and therefore have auto-generated hashCode and equals. No
change of behaviour is expected.

Sidenote: you might be wondering, why did the subclasses use the base
implementation and the auto-generated one? Apparently, this behaviour
is documented in the spec. See this SO answer for details
https://stackoverflow.com/a/44990210/262432.

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen reviewed (approved):
Yes, if you have evidence this is a hotspot, then this does look like a valid fix. Yes the current semantics should be equivalent to what you get from the default implementation.

SparkQA commented:
**[Test build #3936 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3936/testReport)** for PR 19369 at commit [`3b1cb50`](https://github.com/apache/spark/commit/3b1cb50dd1481ee4edbe767916c187709cee2de8).

superbobry commented:
> Yes, if you have evidence this is a hotspot, then this does look like a valid fix. 

I don't think it's a hotspot (otherwise it would have probably been reported long before). I do think however that there is no reason to stick to the old implementation and produce garbage on each `hashCode` call. 

On our side, the issue popped up while profiling GC pauses on the drive during the shuffle. The pauses were caused by an unrelated issue, but we did notice a surprisingly large amount (~10% of all objects on the heap) of `char[]` and `String` for a job which primarily did double operations. Note that the percentage is in terms of the number of objects, not their size.

You can reproduce the issue with a toy example which doesn't do any `String` allocations.

```
val sc = SparkSession.builder()
    .appName("demo")
    .master("local[*]")
    .getOrCreate()
    .sparkContext

val rdd = sc.range(0, 1L << 63, numSlices = 1000).map(_ => ThreadLocalRandom.current.nextLong())
    .map(n => n % 1000 -> n)
    .persist(StorageLevel.DISK_ONLY)
rdd.count()
while (true) {
  rdd.reduceByKey(_ + _).collect()
}
```

Run it and monitor the heap with `jmap`. Here's what I get after a couple of minutes

```
map -histo $PID | head -n 10   

 num     #instances         #bytes  class name
----------------------------------------------
   1:       1895743      137954064  [C
   2:        111418       65757664  [B
   3:        944978       45248496  [Ljava.lang.Object;
   4:           831       27243504  [Lorg.apache.spark.unsafe.memory.MemoryBlock;
   5:         99471       22199824  [I
   6:        865400       20769600  java.lang.String
   7:        761323       18271752  scala.Tuple2                                                                                                                         
```

srowen commented:
I don't know that this is evidence that the allocations came from this bit of code, but, still it seems like a win.

SparkQA commented:
**[Test build #3936 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3936/testReport)** for PR 19369 at commit [`3b1cb50`](https://github.com/apache/spark/commit/3b1cb50dd1481ee4edbe767916c187709cee2de8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

superbobry commented:
Will have a look at the failing tests tomorrow and also have a look at the origins of the strings in the sample above.

Commit:
Changed BlockReplicationUtils.getRandomSample to be monotone

Floyd algorithm is not monotone wrt m. Even if the seed is fixed the
m and m + 1 samples might have less than m elements in common. This
commit replaces the Floyd algorithm with Knuth shuffle, which requires
linear space and time, but guarantees the monotonicity.

superbobry commented:
I've looked at the allocation profile of the sample, it does indeed contain strings from other sources, so the heap histogram is not representative. However, the allocations from `BlockManagerMasterEndpoint` are gone.

The test failures are due to an invalid test "deterministic node selection" in the `BlockManagerReplicationBehavior`. The test checks that any `BlockReplicationPolicy` is deterministic and strictly monotonic, meaning that the locations for 4x replication of some block would be a strict superset of 3x and 2x locations. This is not the case for neither `BasicBlockReplicationPolicy` nor 
 `RandomBlockReplicationPolicy`. Both call `getSampleIds` and, for them to be deterministic given a fixed random seed `r`, the result of `getSampleIds(n, m, r)` should be a subset of `getSampleIds(n, m + 1, r)`. This would be true if the method implemented a partial Knuth shuffle, which does the same sequence of `m` `nextInt` calls in both cases. For Floyd algorithm, however, this is false, because the sequence of calls to `nextInt` depends on `m`. Specifically for `getSampleIds(n, m, r)` it would start from

```
r.nextInt(n - m + 1)
...
```

while for `getSampleIds(n, m + 1, r)` the first call would be

```
r.nextInt(n - (m + 1) + 1)
...
```

Magically hashing `"test_" + id` resulted in the case where the monotonicity property is satisfied. The switch to the auto-generated `hashCode` has broken the magic. 

I replaced the implementation of `getSampleIds` with 

```scala
private def getSampleIds(n: Int, m: Int, r: Random): List[Int] = {
  r.shuffle(List.range(0, n)).take(m)
}
```

Commit:
Changed DiskStore.blockSizes to use BlockId instead of its name

Commit:
Changed NettyBlockTransferService.uploadBlock to use BlockId.name

Previously it relied on toString returning name which clearly was an
abstraction leak.

srowen commented:
Yeah, not surprised that it's due to a test that relies a little too much on how data happens to hash or partition. The fix looks OK.

jiangxb1987 commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -67,7 +67,7 @@ private[spark] class DiskStore(
     try {
       writeFunc(out)
-      blockSizes.put(blockId.name, out.getCount)
+      blockSizes.put(blockId, out.getCount)
```
Though this is equal to the previous version of code, what do we gain from making this change?

superbobry replied:
This is in line with the other part of the PR: `name` allocates a String/StringBuilder. Why use it as a hash key and pay the price of unnecessary allocation, when `BlockId` is hashable?

srowen replied:
@superbobry I think the last test failure is legit as you need to update the call to remove(blockId.name) on about line 116.

I was surprised it even compiles, but, for legacy reasons the JDK collections classes don't have generic types on methods like remove, so it accepts any object. That however should be the last change here.

srowen commented on core/src/main/scala/org/apache/spark/storage/DiskStore.scala:
```diff
@@ -49,7 +49,7 @@ private[spark] class DiskStore(
   private val blockSizes = new ConcurrentHashMap[String, Long]()
 
-  def getSize(blockId: BlockId): Long = blockSizes.get(blockId.name)
+  def getSize(blockId: BlockId): Long = blockSizes.get(blockId)
```
Hm, how does this work if the map keys are still strings?

superbobry replied:
Bah, sorry, fixed.

srowen commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPolicy.scala:
```diff
@@ -85,11 +65,9 @@ object BlockReplicationUtils {
-    } else {
-      r.shuffle(elems).toList
-    }
+    // This takes linear space, but is stable wrt m. That is for a fixed
```
Hm, but I presume the purpose of the specialized method is to handle the case where elems is large. This isn't just test code so this might have a significant impact. Can the existing approach not be modified to be stable, or tests otherwise relaxed?

superbobry replied:
Yes, the purpose was indeed that.

Do you think the contract tested in `BlockManagerReplicationBehavior` makes sense? Personally, I fail to see why this contract is important, but if there is a need I don't think we can do better than O(elems.size). Otherwise, the solution is to just remove the test.

srowen replied:
I see now. I suspect you should just remove the test instead. @shubhamchopra what do you think? I believe you created most of this code.

BTW if performance really mattered here we could make this sampling method a little bit more efficient (avoid foldLeft, -1 at the end, etc) but I don't think it matters

srowen replied:
@superbobry what do you think about just removing the test?

superbobry replied:
Sounds good to me. Will push an update.

Commit:
Revert "Changed BlockReplicationUtils.getRandomSample to be monotone"

This reverts commit 25ef2c2341d091a86cc3995d8f39c823d505d809.

Commit:
Removed "deterministic node selection" test from BlockManagerReplicationSuite

The selection is not fully deterministic, it might yield non-overlapping
node subsets when going from k to (k + 1) replication. See discussion in

srowen reviewed (commented):
CC @tdas because it removes a test you wrote a long time ago. I think I agree with the logic of the change here, that the behavior the test asserts isn't actually needed or guaranteed

SparkQA commented:
**[Test build #3941 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3941/testReport)** for PR 19369 at commit [`8590efe`](https://github.com/apache/spark/commit/8590efec78638735f170e9f6d2fd04c65724e20e).

SparkQA commented:
**[Test build #3941 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3941/testReport)** for PR 19369 at commit [`8590efe`](https://github.com/apache/spark/commit/8590efec78638735f170e9f6d2fd04c65724e20e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

Commit:
Removed leftover BlockId.name usage from DiskStore

superbobry commented:
I've fixed the failing `DiskStoreSuite` and ensured the other two suites also pass fine.

SparkQA commented:
**[Test build #3942 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3942/testReport)** for PR 19369 at commit [`d996c28`](https://github.com/apache/spark/commit/d996c283602269afd05dffad1e681f47f7baf47f).

SparkQA commented:
**[Test build #3942 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3942/testReport)** for PR 19369 at commit [`d996c28`](https://github.com/apache/spark/commit/d996c283602269afd05dffad1e681f47f7baf47f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
Merged to master
